crates/turborepo-lib/src/daemon/server.rs

```diff
@@ -0,0 +1,313 @@
+            timeout,
+
+            watcher,
+            shutdown: Mutex::new(Some(send_shutdown)),
+            shutdown_rx: Some(recv_shutdown),
+        })
+    }
+}
+
+impl<T: Watcher + Send + 'static> DaemonServer<T> {
+    /// Serve the daemon server, while also watching for filesystem changes.
+    pub async fn serve(mut self, repo_root: AbsoluteNormalizedPathBuf) -> CloseReason {
+        let stop = StopSource::new();
+        let watcher = self.watcher.clone();
+        let watcher_fut = watcher.watch(repo_root.to_path_buf(), stop.token());
+
+        let timer = self.timeout_time.clone();
+        let timeout_fut = timer.wait();
+
+        // if shutdown is available, then listen. otherwise just wait forever
+        let shutdown_rx = self.shutdown_rx.take();
+        let shutdown_fut = async move {
+            match shutdown_rx {
+                Some(rx) => {
+                    rx.await.ok();
+                }
+                None => {
+                    futures::pending!();
+                }
+            }
+        };
+
+        #[cfg(feature = "http")]
+        let server_fut = {
+            // set up grpc reflection
+            let efd = include_bytes!("file_descriptor_set.bin");
+            let reflection = tonic_reflection::server::Builder::configure()
+                .register_encoded_file_descriptor_set(efd)
+                .build()
+                .unwrap();
+
+            Server::builder()
+                .add_service(reflection)
+                .add_service(crate::daemon::proto::turbod_server::TurbodServer::new(self))
+                .serve_with_shutdown("127.0.0.1:5000".parse().unwrap(), shutdown_fut)
+        };
+
+        #[cfg(not(feature = "http"))]
+        let (_lock, server_fut) = {
+            let (lock, stream) = crate::daemon::endpoint::open_socket(self.daemon_root.clone())
+                .await
+                .unwrap();
+
+            (
+                lock,
+                Server::builder()
+                    .add_service(crate::daemon::proto::turbod_server::TurbodServer::new(self))
+                    .serve_with_incoming(stream),
+            )
+        };
+
+        select! {
+            _ = server_fut => CloseReason::ServerClosed,
+            _ = watcher_fut => CloseReason::WatcherClosed,
+            _ = shutdown_fut => CloseReason::Shutdown,
+            _ = timeout_fut => CloseReason::Timeout,
+            _ = ctrl_c() => CloseReason::Interrupt,
+        }
+
+        // here the stop token is dropped, and the pid lock is dropped
+        // causing them to be cleaned up
+    }
+
+    fn bump_timeout(&self) {
+        self.timeout_time.reset(self.timeout);
+    }
+}
+
+#[tonic::async_trait]
+impl<T: Watcher + Send + 'static> proto::turbod_server::Turbod for DaemonServer<T> {
+    async fn hello(
+        &self,
+        request: tonic::Request<proto::HelloRequest>,
+    ) -> Result<tonic::Response<proto::HelloResponse>, tonic::Status> {
+        self.bump_timeout();
```

Comment:
Can we use an [interceptor](https://docs.rs/tonic/latest/tonic/service/trait.Interceptor.html) ([example](https://github.com/hyperium/tonic/blob/master/examples/src/interceptor/server.rs#L37)) like in the Go implementation so that we don't have to remember to `bump_timeout` on every call of every service?

Reply:
I have reworked this into a `tower` layer which is compatible with any hyper-based server (not just grpc servers).